src/v/storage/segment_utils.cc

```diff
@@ -741,7 +741,9 @@ ss::future<compaction_result> self_compact_segment(
           "Cannot compact an active segment. cfg:{} - segment:{}", cfg, s));
     }
 
-    if (s->finished_self_compaction() || !s->has_compactible_offsets(cfg)) {
+    if (
+      (s->finished_self_compaction() || !s->has_compactible_offsets(cfg))
+      && !may_have_removable_tombstones(s, cfg)) {
```

Comment:
Very important for allowing self compaction to remove tombstones, even if the segment has already been through a round of self compaction.